bpf/ut/failsafes_test.go

```diff
@@ -0,0 +1,207 @@
+
+package ut_test
+
+import (
+	"net"
+	"testing"
+
+	"github.com/google/gopacket/layers"
+	. "github.com/onsi/gomega"
+
+	"github.com/projectcalico/felix/bpf/failsafes"
+	"github.com/projectcalico/felix/bpf/polprog"
+	"github.com/projectcalico/felix/bpf/routes"
+	"github.com/projectcalico/felix/ip"
+	"github.com/projectcalico/felix/proto"
+)
+
+func TestFailsafes(t *testing.T) {
+	RegisterTestingT(t)
+
+	defer resetBPFMaps()
+
+	hostIP = dstIP // set host IP to the default dest
+	hostCIDR := ip.CIDRFromNetIP(hostIP).(ip.V4CIDR)
+
+	// Setup routing so that failsafe check knows it is localhost
+	rtKey := routes.NewKey(hostCIDR).AsBytes()
+	rtVal := routes.NewValueWithIfIndex(routes.FlagsLocalHost, 1).AsBytes()
+	err := rtMap.Update(rtKey, rtVal)
+	Expect(err).NotTo(HaveOccurred())
+
+	// Set up failsafe to accept incoming connections from srcIP (1.1.1.1/16)
+	err = fsafeMap.Update(
+		failsafes.MakeKey(17, 5678, false, srcIP.String(), 16).ToSlice(),
+		failsafes.Value(),
+	)
+	Expect(err).NotTo(HaveOccurred())
+
+	// Set up failsafe to accept outgoing connections to 3.3.3.3/16
+	fsafeDstIP := net.IPv4(3, 3, 3, 3)
+	err = fsafeMap.Update(
+		failsafes.MakeKey(17, 5678, true, fsafeDstIP.String(), 16).ToSlice(),
+		failsafes.Value(),
+	)
+	Expect(err).NotTo(HaveOccurred())
+
+	denyAllRules := polprog.Rules{
+		ForHostInterface: true,
+		HostNormalTiers: []polprog.Tier{{
+			Policies: []polprog.Policy{{
+				Name: "deny all",
+				Rules: []polprog.Rule{{Rule: &proto.Rule{
+					Action: "Deny",
+				}}},
+			}},
+		}},
+	}
+
+	denyAllRulesWorkloads := polprog.Rules{
+		Tiers: []polprog.Tier{{
+			Policies: []polprog.Policy{{
+				Name: "deny all",
+				Rules: []polprog.Rule{{Rule: &proto.Rule{
+					Action: "Deny",
+				}}},
+			}},
+		}},
+	}
+
+	// Packet from 1.1.1.1 to 2.2.2.2
+	iphdr := *ipv4Default
+	_, _, _, _, pktBytesIn, err := testPacket(nil, &iphdr, nil, nil)
+	Expect(err).NotTo(HaveOccurred())
+	// Packets from failsafe IP to localhost are allowed
+	runBpfTest(t, "calico_from_host_ep", &denyAllRules, func(bpfrun bpfProgRunFn) {
+		res, err := bpfrun(pktBytesIn)
+		Expect(err).NotTo(HaveOccurred())
+		Expect(res.RetvalStr()).To(Equal("TC_ACT_UNSPEC"), "expected program to return TC_ACT_UNSPEC")
+	})
+
+	// Packet from IP not in failsafe (4.4.4.4) to localhost (2.2.2.2)
+	ipHeaderInBlocked := &layers.IPv4{
+		Version:  4,
+		IHL:      5,
+		TTL:      64,
+		Flags:    layers.IPv4DontFragment,
+		SrcIP:    net.IPv4(4, 4, 4, 4),
+		DstIP:    dstIP,
+		Protocol: layers.IPProtocolUDP,
+	}
+	_, _, _, _, pktBytesInBlocked, err := testPacket(nil, ipHeaderInBlocked, nil, nil)
+	Expect(err).NotTo(HaveOccurred())
+	// Packets from non-failsafe IP to localhost are denied
+	runBpfTest(t, "calico_from_host_ep", &denyAllRules, func(bpfrun bpfProgRunFn) {
+		res, err := bpfrun(pktBytesInBlocked)
+		Expect(err).NotTo(HaveOccurred())
+		Expect(res.RetvalStr()).To(Equal("TC_ACT_SHOT"), "expected program to return TC_ACT_SHOT")
+	})
```

Comment:
Since the body of the test is the same, it would be nice to structure the test as a table, in which we would just spec the src/dst ip and ports and expected verdict. It would be easier to read the test cases. Sort of like here https://github.com/projectcalico/felix/blob/master/bpf/ut/pol_prog_test.go#L232

Reply:
Good idea, I'll refactor to make it a bit less verbose